mm2src/mm2_main/src/lp_ordermatch.rs

```diff
@@ -2977,79 +3016,174 @@ fn lp_connect_start_bob(ctx: MmArc, maker_match: MakerMatch, maker_order: MakerO
+                locktime: &lock_time,
+            };
+            start_maker_legacy_swap(&ctx, maker_order, alice, secret, params).await;
+            return;
+        }
+
+        if ctx.use_trading_proto_v2() && alice_swap_v == Some(2) {
+            let params = StateMachineParams {
+                secret_hash_algo: &detect_secret_hash_algo(&maker_coin, &taker_coin),
+                uuid: &uuid,
+                my_conf_settings: &my_conf_settings,
+                locktime: &lock_time,
+                maker_amount: &maker_amount,
+                taker_amount: &taker_amount,
+            };
+            let taker_p2p_pubkey = match taker_p2p_pubkey {
+                PublicKey::Secp256k1(pubkey) => pubkey.into(),
+            };
+
+            match (&maker_coin, &taker_coin) {
                 (MmCoinEnum::UtxoCoin(m), MmCoinEnum::UtxoCoin(t)) => {
-                    let mut maker_swap_state_machine = MakerSwapStateMachine {
-                        storage: MakerSwapStorage::new(ctx.clone()),
-                        abortable_system: ctx
-                            .abortable_system
-                            .create_subsystem()
-                            .expect("create_subsystem should not fail"),
-                        ctx,
-                        started_at: now_sec(),
-                        maker_coin: m.clone(),
-                        maker_volume: maker_amount,
-                        secret,
-                        taker_coin: t.clone(),
-                        dex_fee: dex_fee_amount_from_taker_coin(&t, m.ticker(), &taker_amount),
-                        taker_volume: taker_amount,
-                        taker_premium: Default::default(),
-                        conf_settings: my_conf_settings,
-                        p2p_topic: swap_v2_topic(&uuid),
-                        uuid,
-                        p2p_keypair: maker_order.p2p_privkey.map(SerializableSecp256k1Keypair::into_inner),
-                        secret_hash_algo,
-                        lock_duration: lock_time,
-                        taker_p2p_pubkey: match taker_p2p_pubkey {
-                            PublicKey::Secp256k1(pubkey) => pubkey.into(),
-                        },
+                    start_maker_swap_state_machine(&ctx, &maker_order, &taker_p2p_pubkey, &secret, m, t, &params).await;
+                },
+                (MmCoinEnum::EthCoin(m), MmCoinEnum::EthCoin(t)) => {
+                    start_maker_swap_state_machine(&ctx, &maker_order, &taker_p2p_pubkey, &secret, m, t, &params).await;
+                },
+                (MmCoinEnum::UtxoCoin(m), MmCoinEnum::EthCoin(t)) => {
+                    start_maker_swap_state_machine(&ctx, &maker_order, &taker_p2p_pubkey, &secret, m, t, &params).await;
+                },
+                (MmCoinEnum::EthCoin(m), MmCoinEnum::UtxoCoin(t)) => {
+                    start_maker_swap_state_machine(&ctx, &maker_order, &taker_p2p_pubkey, &secret, m, t, &params).await;
```

Comment:
what about having some method `.supports_v2` and use it in the match to decide which route to go.

Reply:
lemme elaborate:
```rust
match (c1, c2) {
(coin1, coin2) => if coin1.supports_v2() && coin2.supports_v2() { unwrap the coins from the enum & start v2 swap }
(coin1, coin2) => { one (or both) of these doesn't support v2, so start a legacy swap }
}
```
matching against multiple match args just doesn't scale very well

Reply:
> lemme elaborate

yep yep I got it, I left emoji like w00w exactlyyy :D I didnt like such over bloating matching, but obvious solution didnt visit my mind

Reply:
wait a moment, how are you going to "unwrap" two enums in the same arm?

Reply:
discussed over DMs:
this isn't actually possible as the swap state machine accepts generic `MmCoin`s which it needs to know the size of at compile time.
Could be avoided if the swap state machine made with `Arc-/Box<dyn MmCoin>` instead.

Reply:
> discussed over DMs: this isn't actually possible as the swap state machine accepts generic `MmCoin`s which it needs to know the size of at compile time. Could be avoided if the swap state machine made with `Arc-/Box<dyn MmCoin>` instead.

Current state machine requires coins to implement MmCoin + MakerCoinSwapOpsV2 and TakerCoinSwapOpsV2 traits not only MmCoin

Reply:
e.g. with associated types: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c73db6cce48556a388ecafe86a87842d
that was actually hard to nail, needed a dyn refresher. not the best looking though.